metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -92,17 +84,55 @@ class ExtractRenameMember(
 
   }
 
-  private def membersDefinitions(tree: Tree): List[Member] = {
-    val nodes: ListBuffer[Member] = ListBuffer()
+  private def canRenameDefn(
+      defn: Member,
+      fileName: String,
+      defnAtCursor: Option[Member],
+      uri: String
+  ): Boolean = {
+    val differentNames = defn.name.value != fileName
+    val newFileUri = newPathFromClass(uri, defn)
+
+    differentNames && !newFileUri.exists && defnAtCursor.exists(
+      _.equals(defn)
+    )
+  }
+
+  private def canExtractDefn(
+      defn: Member,
+      fileName: String,
+      sealedNames: List[String],
+      definitions: List[Member],
+      uri: String
+  ): Boolean = {
+    val differentNames = defn.name.value != fileName
+    val notExtendsSealedOrSealed = notSealed(defn, sealedNames)
+    val companion = definitions.find(c => {
+      !c.equals(defn) && c.name.value.equals(defn.name.value)
+    })
+    val companionNotSealed =
+      companion.exists(notSealed(_, sealedNames)) || companion.isEmpty
+
+    val newFileUri = newPathFromClass(uri, defn)
+
+    differentNames && notExtendsSealedOrSealed && companionNotSealed && !newFileUri.exists
+  }
+
+  private def membersDefinitions(tree: Tree): List[EndableMember] = {
+    val nodes: ListBuffer[EndableMember] = ListBuffer()
 
     val traverser = new SimpleTraverser {
       override def apply(tree: Tree): Unit = tree match {
         case p: Pkg =>
           super.apply(p)
-        case c: Defn.Class => nodes += c
-        case t: Defn.Trait => nodes += t
-        case o: Defn.Object => nodes += o
-        case e: Defn.Enum => nodes += e
+        case c: Defn.Class => nodes += EndableMember(c, None)
+        case t: Defn.Trait => nodes += EndableMember(t, None)
+        case o: Defn.Object => nodes += EndableMember(o, None)
+        case e: Defn.Enum => nodes += EndableMember(e, None)
+        case endMarker: Term.EndMarker =>
+          val last = nodes.last
+          nodes.remove(nodes.size - 1)
+          nodes += EndableMember(last.member, Some(endMarker))
```

Comment:
`dropRight` returns the rest of the list, not the last element! since `.last` could through exception, I replaced it with `findLast(_ => true)` which is safe!

Reply:
I tried that as well as `lastOption`! Anything but the existing code does not work for some reason!